### DISCUSSION TITLE
ActionHandlers Client to Server

### DISCUSSION THREAD

**Discussion by JValGLZ** [eclipse-glsp/glsp#1252] [General] [answered]
Hello, I am currently testing with actions and action handlers. I am using the Java-EMF-Theia template. As of now, I have a working handler that gets user input and stores it inside the action through the client. The next part is, that I want to send this to the server so that it can analyze the expression given by the user and return a result. I understand that I must make an action in the server that will match the action made in the client. I am curious about how it is that from my client's action handler, I can send it to the server. Below is my action handler in the client. Thank you.

```ts
@injectable()
export class MyTestResponseActionHandler implements IActionHandler {
   handle(action: MyTestActionVersionTwo): void | Action {
        // implement your custom logic to handle the action
        console.log('I printed this from the client handler');
        console.log('The text within the input is: ' + action.variable1);
        // Optionally issue a response action
    }
}```

**Comment by tortmayr** [comment, marked as answer]
Hi @JValGLZ,

>As of now, I have a working handler that gets user input and stores it inside the action through the client. The next part is, that I want to send this to the server so that it can analyze the expression given by the user and return a result

I would recommend to split this into two actions, let's call them `MyTestActionVersionTwo` and `MyTestActionVersionTwoResult` for now.
On the client you than handle the `MyTestActionVersionTwo` and derive the `MyTestActionVersionTwoResult` from that (stores the user input) and then you send this action to the server. 

On the server you have to register a handler for the result action:

```java
// Define your action
public class MyTestActionVersionTwoResult extends Action {
    public static final String KIND="myTestActionVersionTwoResult";
    
    // define the custom properties + getter&setter for your action
}

// Define an action handler
public class MyTestActionVersionTwoResultHandler extends AbstractActionHandler<MyTestActionVersionTwoResult> {
   @Override
   public List<Action> executeAction(final MyTestActionVersionTwoResult action) {
     // Implement your custom handling
     
     // If you want to send back an action to the client use `listOf()` otherwise simple return an empty list with `none()`
     return none();
   }
}

// Register your handler in your custom diagram module
public class TasklistDiagramModule extends EmfNotationDiagramModule {
 //...
    @Override
   protected void c void configureActionHandlers(final MultiBinding<ActionHandler> binding) {
      super.configureActionHandlers(binding);
      binding.add(MyTestActionVersionTwoResultHandler.class);
   }
}
```
In you client action dispatcher you can then simply return the result action, and the client will take care of forwarding it to the server.

```ts
@injectable()
export class MyTestResponseActionHandler implements IActionHandler {

   handle(action: MyTestActionVersionTwo): void | Action {
        // implement your custom logic to handle the action
        console.log('I printed this from the client handler');
        console.log('The text within the input is: ' + action.variable1);
        // dervive the result action
        const result:MyTestActionVersionTwo = ...
        return result:
    }
}```

**Reply by JValGLZ** [reply]
Hello, I wanted to know what exactly would be present within this line `const result:MyTestActionVersionTwo = ...` if I only plan to send the expression and receive the result from the server.

Do I create another action within the client with the same name as the action in the server called MyTestActionVersionTwoResult?

How is it that the client will send that action to the server. 

Also is there a way to monitor the print statements from within the server. I am trying to print out statements to monitor steps, but cannot get them to appear on the terminal. 

Thank you.

**Reply by tortmayr** [reply]
>Do I create another action within the client with the same name as the action in the server called MyTestActionVersionTwoResult?

Yes, exactly you create an action definition on the client that mirrors the action definition of the server.

>How is it that the client will send that action to the server.

As part of the initial handshake between client and server on diagram startup the exchange a list of actions that they can handle.
Meaning the server tells the client which action kinds it can handle (i.e. has handlers for) and vice versa. Whenever an action with a server handled action kind is dispatched on the client it will automatically forward it to the client.

>Also is there a way to monitor the print statements from within the server. I am trying to print out statements to monitor steps, but cannot get them to appear on the terminal.

The `java-emf-theia` template per default only has file logging activated for the GLSP server. Server logs will be stored in `glsp-client/tasklist-theia/logs`.
You can customize this in the [`TaskListGLSPServerContribution`](https://github.com/eclipse-glsp/glsp-examples/blob/1a4484fb573295d60210808f27ee4a315142cd15/project-templates/java-emf-theia/glsp-client/tasklist-theia/src/node/glsp-server-contribution.ts#L36) by setting the `--consoleLog` flag to true in the `additionalArgs`.

**Reply by JValGLZ** [reply]
Thank you so much for your help! I managed to send the information to the server and retrieve it back to the client. I do have a question, even when I set the flag to true, I still cannot see the outputs from within the server. As of now, I tested that the correct data was retrieved by sending it from the client to the server, which changed the values, then back to the client and then printing it. I would still like to print from the server as I continue implementing the handlers. I do the outputs on the java server with System.out.println(), I am not sure if there is another way or if that is the issue. Thank you.
